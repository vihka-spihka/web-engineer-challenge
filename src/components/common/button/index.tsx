import React from 'react';

interface IButton {
	label: string;
	onClick: (e: React.SyntheticEvent) => void;
};

export const Button = ({ label, onClick }: IButton) => {
	return (
		<button onClick={() => onClick}>
			{label}
		</button>
	);
};
