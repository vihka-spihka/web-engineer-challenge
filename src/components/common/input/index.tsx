import React from 'react';
import { StyledInput } from './../styles';

interface IInput extends React.InputHTMLAttributes<HTMLInputElement> {
	label: string;
	onChangeInput: (v: InputValue) => void;
};

export const Input: React.FC<IInput> = ({ label, value, onChangeInput, ...props }) => {
	return (
		<StyledInput>
			<label>
				{label}
				<input
					value={value}
					onChange={
						(e: React.ChangeEvent<HTMLInputElement>) => onChangeInput(e.target.value)
					}
					{...props} />
			</label>
		</StyledInput>
	);
};
