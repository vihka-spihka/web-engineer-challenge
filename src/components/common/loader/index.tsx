import React from 'react';
import { StyledLoader } from './../styles';

export const Loader: React.FC = () => {
	return (
		<StyledLoader>
			<img src='/favicon.ico' width='80' height='80' alt='Company logo' />
		</StyledLoader>
	);
};
