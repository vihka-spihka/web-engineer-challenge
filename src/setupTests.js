// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { ThemeProvider } from 'styled-components';

export const defaultTheme = {
	data: {
		global: {
			body_color: 'green',
			loading_bar_color: 'red',
			background_color: 'white',
			primary_border_color: 'black'
		},
		primary_font_family: {
			regular: 'Font',
			bold: 'Font',
		},
		buttons: {
			primary_button_text_color: 'blue',
			primary_button_font_family: 'blue',
			primary_button_border_color: 'blue',
			primary_button_border_radius: 10,
			primary_button_background_color: 'blue',
			secondary_button_text_color: 'blue',
			secondary_button_font_family: 'blue',
			secondary_button_border_color: 'blue',
			secondary_button_background_color: 'blue',
			secondary_button_border_radius: 10,
			knockout_button_text_color: 'blue',
			knockout_button_font_family: 'blue',
			knockout_button_border_color: 'blue',
			knockout_button_background_color: 'blue',
			knockout_button_border_radius: 10,
		},
		forms: {
			field_height: 50,
			placeholder_color: 'black',
		},
		header: {
			font_color: 'blue',
			border_color: 'red',
			border_radius: 9,
			background_color: 'green',
			desktop_logo_width: 200,
			font_size: 10,
		}
	},
};

/* see issues:
	- https://github.com/styled-components/jest-styled-components/issues/119
	- https://github.com/styled-components/styled-components/issues/1319
*/
export function shallowWithTheme(child) {
	return shallow(child, {
		wrappingComponent: ({ children }) => {
			return (<ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>);
		},
	});
}

Enzyme.configure({ adapter: new Adapter() });
