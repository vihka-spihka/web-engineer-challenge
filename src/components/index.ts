import { Input, StyledButton, Loader } from './common';
import { Countdown } from './countdown';
import { Footer } from './footer';
import { Header } from './header';
import { History } from './history';

export { Input, StyledButton, Loader, Countdown, Footer, Header, History };
