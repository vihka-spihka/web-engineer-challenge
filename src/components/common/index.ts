import { Input } from './input';
import { StyledButton } from './styles';
import { Loader } from './loader';

export { Input, StyledButton , Loader};
