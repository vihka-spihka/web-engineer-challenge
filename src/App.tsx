import React from 'react';
import { ThemeProvider } from 'styled-components';
import { Countdown } from './components/countdown';

interface IState {
  theme: Theme | null;
  countdown: Countdown | null;
}

/* TODO: think about the implementation of IProps as { defaultTheme: ITheme } */
class App extends React.Component<unknown, IState> {
  state: IState = {
  	theme: null,
  	countdown: null,
  }

  componentDidMount() {
  	fetch('https://api.koala.io/marketing/v1/device-configurations/alias/web-config', {
  		method: 'GET',
  		headers: {
  			'X-Organization-Id': '1',
  		}
  	})
  		.then(res => res.json())
  		.then(result => this.setState({ theme: result.data }))
  		.catch(error => console.log(error.message));

  }

  render() {
  	const data = this.state.theme && this.state.theme.data;
  	return (
  		!!this.state.theme && (
  			<ThemeProvider theme={data}>
  				<h1>Countdown Clock</h1>
  				<img src='koala-logo.png' alt='Company logo' />
  				{ this.state.countdown
  					? <div>Seconds left: {this.state.countdown.timeLeft}</div>
  					: <Countdown setCountdown={this.setTimer} />}
  			</ThemeProvider>
  		)
  	);
  }

  setTimer = (time: InputValue): void => {
  	const countdown = setInterval(() => {
  		!this.state.countdown
  			? this.setState({
  				countdown: {
  					id: countdown,
  					start: Date.now(),
  					duration: time as number,
  					timeLeft: time as number,
  				}
  			})
  			: this.setState({
  				countdown: {
  					...this.state.countdown,
  					timeLeft: this.state.countdown.timeLeft - 1
  				}});

  		if (this.state.countdown?.timeLeft === 0) {
  			clearInterval(countdown);
  			this.setState({ countdown: null });
  		}
  	}, 1000);
  };
}

export default App;
