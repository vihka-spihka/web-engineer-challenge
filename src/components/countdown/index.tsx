import React, { useState } from 'react';
import { StyledCountdown } from './styles';
import { Input, Button } from './../common';
import { COUNTDOWN_MODE } from './../../App';

interface IProp {
	mode: COUNTDOWN_MODE;
	timeLeft?: number;
	changeMode: (mode: COUNTDOWN_MODE, time?: number) => void;
};

export const Countdown = ({ mode, timeLeft, changeMode }: IProp) => {
	const [value, setValue] = useState<InputValue>('');

	const submitForm = (e: React.SyntheticEvent): void => {
		e.preventDefault();
		changeMode(COUNTDOWN_MODE.IN_PROGRESS, value as number);
		setValue('');
	};

	const editCountdown = () => {
		timeLeft && setValue(timeLeft);
		changeMode(COUNTDOWN_MODE.EDIT);
	};

	const resetCountdown = (): void => {
		setValue('');
		changeMode(COUNTDOWN_MODE.NEW);
	};

	let countdown: React.ReactElement;

	switch(mode) {
		case COUNTDOWN_MODE.EDIT:
			countdown = (
				<form onSubmit={(e: React.SyntheticEvent) => submitForm(e)}>
					<Input
						type={'number'}
						value={value?.toString()}
						label={'Enter time in seconds: '}
						onChangeInput={val => setValue(val)}
						placeholder={'e.g. 1000'} />
					<Button label={'Done'} onClick={(e: React.SyntheticEvent) => submitForm(e)} />
					<Button label={'Reset'} onClick={() => resetCountdown()} />
				</form>
			);
			break;
		case COUNTDOWN_MODE.IN_PROGRESS:
			countdown = (
				<>
					<div>Seconds left: {timeLeft}</div>
					<Button label={'Edit'} onClick={() => editCountdown()} />
					<Button label={'Reset'} onClick={() => resetCountdown()} />
				</>
			);
			break;
		case COUNTDOWN_MODE.NEW:
		default:
			countdown = (
				<form onSubmit={(e) => submitForm(e)}>
					<Input
						type={'number'}
						value={value?.toString()}
						label={'Enter time in seconds: '}
						onChangeInput={val => setValue(val)}
						placeholder={'e.g. 1000'} />
					<Button label={'Start'} onClick={(e: React.SyntheticEvent) => submitForm(e)} />
				</form>
			);
			break;
	}

	return (
		<StyledCountdown>
			{countdown}
		</StyledCountdown>
	);
};
