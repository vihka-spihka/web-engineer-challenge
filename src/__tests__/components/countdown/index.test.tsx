import React from 'react';
import { ShallowWrapper } from 'enzyme';
import { Countdown } from './../../../components/countdown';
import { COUNTDOWN_MODE } from './../../../App';
import { Input , StyledButton} from './../../../components/common';
import { shallowWithTheme } from './../../../setupTests';

let wrapper: ShallowWrapper;

const useCountdown = (
	mode: COUNTDOWN_MODE,
	timeLeft?: number
): React.ComponentElement<any, any> => {
	return (<Countdown mode={mode} timeLeft={timeLeft} changeMode={jest.fn()} />);
};

const useInput = (
	value: string,
	label: string
): React.ComponentElement<any, any> => {
	return (
		<Input
			type={'number'}
			value={value}
			label={label}
			onChangeInput={expect.any(Function)}
			placeholder={'e.g. 1000'} />
	);
};

const useBtn = (
	label: string,
	style: 'primary' | 'secondary' | 'knockout',
	disabled?: boolean
): React.ComponentElement<any, any> => {
	return (
		<StyledButton
			label={label}
			$styleType={style}
			disabled={disabled}
			onClick={expect.any(Function)} />
	);
};

describe('Countdown', () => {
	it('renders correctly', () => {
		wrapper = shallowWithTheme(useCountdown(COUNTDOWN_MODE.NEW));
		expect(wrapper).toMatchSnapshot();
	});

	it('should show empty input when mode is set to NEW', () => {
		wrapper = shallowWithTheme(useCountdown(COUNTDOWN_MODE.NEW));

		const btn = useBtn('Start', 'knockout', true);
		const countdownInput = useInput('', 'Enter time in seconds: ');

		expect(wrapper.find('form').children().first().getElement()).toEqual(countdownInput);
		expect(wrapper.find('form').children().last().getElement()).toEqual(btn);
	});

	it('should show `Seconds left: 4` when mode is set to IN_PROGRESS', () => {
		wrapper = shallowWithTheme(useCountdown(COUNTDOWN_MODE.IN_PROGRESS, 4));
		const label = 'Seconds left: 4';
		const editBtn = useBtn('Edit', 'primary');
		const resetBtn = useBtn('Reset', 'secondary');

		expect(wrapper.children().first().text()).toEqual(label);
		expect(wrapper.children().last().children().first().getElement()).toEqual(editBtn);
		expect(wrapper.children().last().children().last().getElement()).toEqual(resetBtn);
	});

	it('should show input with value when mode is set to EDIT', () => {
		wrapper = shallowWithTheme(useCountdown(COUNTDOWN_MODE.EDIT, 4));
		
		const doneBtn = useBtn('Done', 'primary', false);
		const resetBtn = useBtn('Reset', 'secondary');
		const countdownInput = useInput('4', 'Enter time in seconds: ');

		expect(wrapper.find('form').childAt(0).getElement()).toEqual(countdownInput);
		expect(wrapper.find('form').childAt(1).getElement()).toEqual(doneBtn);
		expect(wrapper.find('form').childAt(2).getElement()).toEqual(resetBtn);
	});

	afterEach(() => {
		wrapper.unmount();
	});
});
