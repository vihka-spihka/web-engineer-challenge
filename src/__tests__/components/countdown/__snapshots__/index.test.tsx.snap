// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countdown renders correctly 1`] = `
<styled.div>
  <form
    onSubmit={[Function]}
  >
    <Input
      label="Enter time in seconds: "
      onChangeInput={[Function]}
      placeholder="e.g. 1000"
      type="number"
      value=""
    />
    <Styled(Button)
      $styleType="knockout"
      disabled={true}
      label="Start"
      onClick={[Function]}
    />
  </form>
</styled.div>
`;
